### PR TITLE
WP CLI Refactoring

### PR DESCRIPTION
**Pull request by crstauf**
PR for #265 and #272.

### CODE REVIEW

**Comment by rrennick** [comment]
Updated the base for this PR to `version_3_1_0`.

**Comment by crstauf** [comment]
@thenbrent @rrennick Submitted a rebuild from `version_3_2_0`. 

Took me awhile, but I think I got it figured. :joy:

**Comment by crstauf** [comment]
Build failed due to [timestamp `assertEquals()` failure](https://travis-ci.org/github/woocommerce/action-scheduler/jobs/694753302#L186).

**Comment by rrennick** [comment]
> Build failed due to timestamp assertEquals() failure.

@crstauf I triggered a re-run. Whenever the tests fail I always check them to see why and trigger a re-run when necessary.

**Comment by rrennick** [comment]
@crstauf This looks like an interesting refactor if AS were a private in-house project. However, responsible public projects require `public` and `protected` class methods and properties be deprecated (and left functional if possible) before being removed. Let us know if you would like to update this with those restraints in mind.

**Comment by crstauf** [comment]
@rrennick Does look like I removed the `register()` and `migrate()` functions from `Migration_Command`; I'll get those restored to properly use the new methods.

Were there any other functions you saw that were removed without deprecation?

**Comment by rrennick** [comment]
@crstauf If you are going to continue on it then there are some other things that we'd like to see

- Use `git mv` instead of `mv` to move files in the repo. A `mv` is a file delete and file add. `git mv` will just show the changes to the file.
- AS has an autoloader which should be updated instead of using `require()`.
- Abstract classes go in the `/classes/abstracts/` folder.
- Where possible, new WP_CLI classes should be created in the `ActionScheduler\WP_CLI` namespace

**Comment by crstauf** [comment]
> Use `git mv` instead of `mv`

Will check on that.

> AS has an autoloader which should be updated instead of using `require()`.

This was discussed previously and determined okay so long as was in main plugin file: https://github.com/woocommerce/action-scheduler/pull/279#discussion_r271825153. Please indicate if this policy has changed and is no longer acceptable.

> Abstract classes go in the `/classes/abstracts/` folder.

I think there was something that wasn't working when I tried in the `abstracts` folder... perhaps it was as simple as I wasn't sure which folder it should go in (`WP_CLI` or `abstracts`).

> new WP_CLI classes should be created in the `ActionScheduler\WP_CLI` namespace.

I'm not 100% comfortable working with namespaces yet. I'll give it a go, but may need some support.

Thanks, @rrennick!

**Comment by rrennick** [comment]
> Will check on that.

The best way to do that is create a new branch, move the files then apply the changes. Once you have rebuilt the PR, rename the branches and push the new branch you created.

**Comment by crstauf** [comment]
@rrennick I've removed the duplicate file, and I'm not sure about your other comments.

> Use `git mv` instead of `mv`

The only file that was moved is `ActionScheduler_WPCLI_Scheduler_command`, and everything in the file was changed, so nearly the entire file is new. Is that the file you are referring to?

> AS has an autoloader which should be used instead of using `require()`.

This was discussed previously and determined okay so long as was in main plugin file: #279 (comment). Please indicate if this policy has changed and is no longer acceptable.

> new WP_CLI classes should be created in the `ActionScheduler\WP_CLI` namespace.

All WP CLI commands should already be in the namespace. Please confirm?

**Comment by rrennick** [comment]
> Is that the file you are referring to?

Yes, when you git move the file you retain the history between the new and old file.

> This was discussed previously and determined okay so long as was in main plugin file

The main plugin file is `/action-scheduler/action-scheduler.php` vs the `ActionScheduler` class.

> All WP CLI commands should already be in the namespace.

There were no namespaces in the plugin prior to 3.0 and 3.0+ has not moved any non-namespaced classes to namespaces (eg. https://github.com/woocommerce/action-scheduler/blob/master/classes/WP_CLI/ActionScheduler_WPCLI_QueueRunner.php).

**Comment by crstauf** [comment]
> retain the history between the new and old file.

@rrennick Thank you for confirming; I'll give that a try.

> main plugin file is `/action-scheduler/action-scheduler.php`

Will attempt to fix.

> there were no namespaces in the plugin prior to 3.0

Both `Migration_Command.php` and `Run_Command.php` have `namespace Action_Scheduler\WP_CLI;` at the top of the file. Is there something else I need to do to enter them into the namespace?

**Comment by rrennick** [comment]
> Both Migration_Command.php and Run_Command.php have namespace Action_Scheduler\WP_CLI;

That's how you add a class to a namespace https://www.php.net/manual/en/language.namespaces.php

**Comment by crstauf** [comment]
> That's how you add a class to a namespace

@rrennick Thanks, so what else needs to be done per [your comment](https://github.com/woocommerce/action-scheduler/pull/280#issuecomment-655172836)?

File: classes/WP_CLI/ActionScheduler_WPCLI.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
+<?php
+
```
Comment (rrennick) [inline classes/WP_CLI/ActionScheduler_WPCLI.php:2]: missing namespace

Comment (crstauf) [reply classes/WP_CLI/ActionScheduler_WPCLI.php:2]: Oh, I was checking commands. Will get that namespaced as well.

**Comment by crstauf** [comment]
@rrennick I attempted `git mv`, but the history does not seem to be carrying over.

```
$ git mv "classes/WP_CLI/ActionScheduler_WPCLI_Scheduler_command.php" "deprecated/ActionScheduler_WPCLI_Scheduler_command.php"
```

[History of original](https://github.com/woocommerce/action-scheduler/commits/master/classes/WP_CLI/ActionScheduler_WPCLI_Scheduler_command.php) vs [history of moved](https://github.com/crstauf/action-scheduler/commits/temp/deprecated/ActionScheduler_WPCLI_Scheduler_command.php).

**Comment by rrennick** [comment]
> I attempted git mv, but the history does not seem to be carrying over.

@crstauf That's why you need to create a new branch and follow the process I outlined in https://github.com/woocommerce/action-scheduler/pull/280#issuecomment-655172836

**Comment by crstauf** [comment]
> create a new branch

@rrennick I did... created `temp` branch from `version_3_2_0`, and performed `git mv` command:

```
$ git mv "classes/WP_CLI/ActionScheduler_WPCLI_Scheduler_command.php" "deprecated/ActionScheduler_WPCLI_Scheduler_command.php"
```

**Comment by rrennick** [comment]
Sorry for the confusion. The commit shows that the file was moved so in another year someone looking at the history can see it was moved and go to the previous commit in the changelog and access the history.

**Comment by crstauf** [comment]
> commit shows that the file was moved

@rrennick Ahh, okay, thanks.

**Comment by crstauf** [comment]
> Once you have rebuilt the PR, rename the branches and push the new branch you created.

@rrennick I renamed `issue_265` to `issue_265_old`, and created a new `issue_265` from `version_3_2_0`. I need to force push `issue_265` now, correct?

**Comment by rrennick** [comment]
> I need to force push issue_265 now, correct?

Correct. Some of the discussion may show as outdated but is still accessible.

**Comment by crstauf** [comment]
@rrennick Done. Will you please review the moved file and confirm done properly?

**Comment by crstauf** [comment]
Closing in favor of https://github.com/crstauf/action-scheduler-cli.

**Pull request closed without merging**